source/base/hdf5.cc

```diff
@@ -0,0 +1,1513 @@
+      std::is_same<Container<number>, Vector<number>>::value,
+      const void *>::type
+    get_container_const_pointer(const Container<number> &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong.
+      // Use the first element of FullMatrix to get the pointer to the raw data
+      return &*data.begin();
+    }
+
+
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, FullMatrix<number>>::value,
+      const void *>::type
+    get_container_const_pointer(const Container<number> &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong.
+      // Use the first element of FullMatrix to get the pointer to the raw data
+      return &data[0][0];
+    }
+
+    // This function initializes a container of number type
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, std::vector<number>>::value,
+      Container<number>>::type
+    initialize_container(std::vector<hsize_t> dimensions)
+    {
+      return Container<number>(std::accumulate(
+        dimensions.begin(), dimensions.end(), 1, std::multiplies<int>()));
+    }
+
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, Vector<number>>::value,
+      Container<number>>::type
+    initialize_container(std::vector<hsize_t> dimensions)
+    {
+      return Container<number>(std::accumulate(
+        dimensions.begin(), dimensions.end(), 1, std::multiplies<int>()));
+    }
+
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, FullMatrix<number>>::value,
+      Container<number>>::type
+    initialize_container(std::vector<hsize_t> dimensions)
+    {
+      return FullMatrix<number>(dimensions[0], dimensions[1]);
+    }
+
+    // Convert a HDF5 no_collective_cause code to a human readable string
+    std::string
+    no_collective_cause_to_string(uint32_t no_collective_cause)
+    {
+      std::string message;
+      // The first is not a bitmask compararison, the rest are bitmask
+      // comparisons.
+      // https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-GetMpioNoCollectiveCause
+      // See H5Ppublic.h
+      // Hex codes are used because the HDF5 Group can deprecate some of the
+      // enum codes. For example the enum code H5D_MPIO_FILTERS is not defined
+      // in 1.10.2 because it is possible to use compressed datasets with the
+      // MPI/IO driver.
+
+      // H5D_MPIO_COLLECTIVE
+      if (no_collective_cause == 0x00)
+        {
+          if (message.length() > 0)
+            {
+              message += ", ";
+            }
+          message += "H5D_MPIO_COLLECTIVE";
```

Comment:
Instead of repeating the four lines here, you can do this more easily:
```
std::string message;
auto append_to_message = [message](const char *p) {
  if (message.length() > 0)
    message += ", ";
  message += p;
}
```
and then in each of these places all you have to do is
```
  if (no_collective_clause == 0x00)
    append_to_message ("H5D...");
  else if (...
```

Reply:
Done